src/subtle/reaction.ts

```diff
@@ -0,0 +1,55 @@
+import { Signal } from "signal-polyfill";
+
+/**
+ * Reactions are a way to observe a value and run an effect when it changes.
+ *
+ * The `data` function is run and tracked in a computed signal. It returns a
+ * value that is compared to the previous value. If the value changes, the
+ * `effect` function is called with the new value and the previous value.
+ *
+ * @param data A function that returns the value to observe.
+ * @param effect A function that is called when the value changes.
+ * @param equals A function that compares two values for equality.
+ * @returns A function that stops the reaction.
+ */
+export const reaction = <T>(
+  data: () => T,
+  effect: (value: T, previousValue: T) => void,
+  equals = Object.is
+) => {
+  // Passing equals here doesn't seem to dedupe the effect calls.
+  const computed: Signal.Computed<T> | undefined = new Signal.Computed(data, {
+    equals,
+  });
+  let previousValue = computed.get();
+  let notify: (() => Promise<void>) | undefined = async () => {
+    await 0;
+    // Check if this reaction was unsubscribed
+    if (notify === undefined) {
+      return;
+    }
+    const value = computed.get();
+    if (!equals(value, previousValue)) {
+      try {
+        effect(value, previousValue);
+      } catch (e) {
+        // TODO: we actually want this to be unhandled, but Vitest complains.
```

Comment:
how does vitest complain? we can use expect / assert throws:
- https://vitest.dev/api/assert.html#throws
- https://vitest.dev/api/expect.html#tothrowerror

Reply:
This isn't about asserting that something throws, because the `reaction()` call isn't on the stack anymore by the time the test effect throws, and doesn't return a promise or something that can contain an async error. This is about user code that throws, but since it's called by a notify callback, letting that be uncaught so that uncaught error handling could handle it.

Without the catch(), Vitest produces an error like:
```
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Rejection ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error: Oops
 ❯ effect tests/subtle/reaction.test.ts:124:15
    122|       if (value === 1) {
    123|         thrown = true;
    124|         throw new Error("Oops");
       |               ^
    125|       }
    126|       thrown = false;
 ❯ notify src/subtle/reaction.ts:34:9

This error originated in "tests/subtle/reaction.test.ts" test file. It doesn't mean the error was thrown inside the file itself, but while it was running.
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
```

Ideally, we exclude just this test from Vitest's unhandled error check. If another test, say a users' own tests, generated an unhandled error, we _would_ want Vitest to fail.

I tried adding an uncaughException handler, but that didn't prevent Vitest's from running.

Reply:
hmm -- ok, thanks for explaining!  I'll take a poke after a readme entry is added -- I want to understand this more :thinking:

Reply:
I figured something out for handling this, and providing the erroring behavior to users -- lemme know what you think